### PR TITLE
Query-frontend: Don't retry on HTTP 503 from ingester limiting

### PR DESCRIPTION
**Pull request by aknuds1**
#### What this PR does
Modify the query-frontend to not retry queries that fail with HTTP 503 (Service Unavailable). This error is typically returned from the ingester when it's overloaded, and retrying will only make things worse.

TODO:

- [x] Test in cloud

##### Caveat

@dimitarvdimitrov says the ingester can return HTTP 503 when it’s starting, in which case the query-frontend should still retry. Therefore, we'll probably need to include metadata in the error about the reason (ingester limiting).

##### Update

I have found from testing though that the ingester doesn't in fact return HTTP 503 when starting, it returns the gRPC error code `UNAVAILABLE`. This reaches the query-frontend as HTTP 500 (translated by Prometheus most likely). When discussing with @duricanikolic and @pstibrany we've reached the conclusion that returning HTTP-gRPC 503 from the ingester, when it limits read requests, should be an OK interim solution, since it's discernible from other errors.

#### Which issue(s) this PR fixes or relates to
Fixes https://github.com/grafana/mimir/issues/6000.

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by dimitarvdimitrov — COMMENTED** [review]
is it possible to narrow this down only to when the ingester is deliberately returning 503 due to read path rate-limiting? I'm thinking about some query failing because the ingester was just starting, which I think we want to retry https://github.com/grafana/mimir/blob/8e9160e0d950334d2bde7d14ecc2bab167e1827f/pkg/ingester/ingester.go#L3072

**Comment by aknuds1** [comment]
@dimitarvdimitrov thanks, I added a TODO.

**Comment by 56quarters** [comment]
This feels like a situation where gRPC error codes would give us the granularity required:

* `codes.Unavailable` during startup which should be retried
* `codes.ResourceExhausted` for hitting CPU/Memory limits

As an aside, this problem seems like the intersection of both @duricanikolic 's work on cleaning up error handling in ingesters (#6019) and circuit breaking in the ingester client (#5951). I'm not sure what the correct solution is here but I think it'd help to take a step back and think about how we want this to work holistically.

**Comment by aknuds1** [comment]
@56quarters thanks. Trying gRPC status code `ResourceExhausted`, though, gets translated to HTTP 500 by the time it reaches the query-frontend (`prometheusCodec.DecodeResponse`). I suspect it's because `ResourceExhausted` is not recognized by Prometheus. Might you know how to work around this?

**Comment by duricanikolic** [comment]
Could you please replace
```
	tooBusyError = httpgrpc.Errorf(http.StatusServiceUnavailable,
		"the ingester is currently too busy to process queries, try again later")
```
with
```
	tooBusyError = newErrorWithStatus(
		safeToWrapError("the ingester is currently too busy to process queries, try again later"),
		http.StatusServiceUnavailable,
	)
```
in [ingester/errors.go](https://github.com/grafana/mimir/blob/main/pkg/ingester/errors.go#L28-L29)?

File: pkg/frontend/querymiddleware/codec.go [hunk 1]
```diff
@@ -353,15 +353,20 @@ func (c prometheusCodec) EncodeRequest(ctx context.Context, r Request) (*http.Re
...
 
 func (c prometheusCodec) DecodeResponse(ctx context.Context, r *http.Response, _ Request, logger log.Logger) (Response, error) {
-	if r.StatusCode/100 == 5 {
-		return nil, httpgrpc.ErrorFromHTTPResponse(&httpgrpc.HTTPResponse{
-			Code: int32(r.StatusCode),
-			Body: mustReadResponseBody(r),
-		})
-	} else if r.StatusCode == http.StatusTooManyRequests {
+	switch r.StatusCode {
+	case http.StatusServiceUnavailable:
```
Comment (duricanikolic) [inline pkg/frontend/querymiddleware/codec.go:357]: Could you please add a unit test for this in codec_test.go?

**Review by duricanikolic — CHANGES_REQUESTED** [review]

File: pkg/frontend/querymiddleware/codec.go (hunk 1 above)
Comment (aknuds1) [reply pkg/frontend/querymiddleware/codec.go:357]: Done, PTAL.

**Review by pstibrany — APPROVED** [review]
Thank you.

File: pkg/frontend/querymiddleware/codec.go (hunk 1 above)
Comment (duricanikolic) [reply pkg/frontend/querymiddleware/codec.go:357]: Thank you

File: pkg/api/error/error.go [hunk 2]
```diff
@@ -61,6 +61,8 @@ func (e *apiError) statusCode() int {
 		return http.StatusRequestEntityTooLarge
 	case TypeNotAcceptable:
 		return http.StatusNotAcceptable
+	case TypeUnavailable:
```
Comment (duricanikolic) [inline pkg/api/error/error.go:64]: @56quarters are we sure that this will not introduce any side-effect to the circuit breaker?

Comment (56quarters) [reply pkg/api/error/error.go:64]: Circuit breakers work entirely on turning errors into gRPC codes: https://github.com/grafana/mimir/blob/da98104bb2c606cf4d0c8fd303d73aa82f77b5d8/pkg/ingester/client/circuitbreaker.go#L94-L104

File: CHANGELOG.md [hunk 3]
```diff
@@ -15,6 +15,7 @@
 * [ENHANCEMENT] Distributor: add experimental support for storing metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingesting metrics via OTLP. Enable with `-distributor.enable-otlp-metadata-storage=true`. #5693
 * [ENHANCEMENT] Ingester: added support for sampling errors, which can be enabled by setting `-ingester.error-sample-rate`. This way each error will be logged once in the configured number of times. All the discarded samples will still be tracked by the `cortex_discarded_samples_total` metric. #5584 #6014
 * [ENHANCEMENT] Ruler: Fetch secrets used to configure TLS on the Alertmanager client from Vault when `-vault.enabled` is true. #5239
+* [BUGFIX] Query-frontend: Don't retry read requests rejected by the ingester due to limiting. #6032
```
Comment (dimitarvdimitrov) [inline CHANGELOG.md:18]: i'd clarify which limiting was broken

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM, only a nitpick about the changelog entry

File: CHANGELOG.md (hunk 3 above)
Comment (aknuds1) [reply CHANGELOG.md]: Done.

**Review by duricanikolic — APPROVED** [review]

**Comment by aknuds1** [comment]
I've successfully tested in cloud, could see from logs (or the absence thereof) that the query-frontend doesn't retry the queries rejected by ingesters due to utilization based limiting.

**Comment by aknuds1** [comment]
I could also see from traces that with my fix, there is only one try in the query-frontend, and two without my fix.

Without fix:
Screen Shot 2023-09-20 at 11 18 49

Before fix:
Screen Shot 2023-09-20 at 11 11 56

**Review by colega — COMMENTED** [review]
I see this is already merged but:

- I think we should not send HTTP status codes in the gRPC Status field by just casting the type. This only makes code harder to understand and creates further bugs.
- This needs an end-to-end test of some kind: either starting some dockers or just starting the components in a single test. I see a couple of changes here, and a couple of changes there, but I think it's hard to say for sure these changes will cover the problem, as there's a lot of components in between (PromQL engine) that can change the error mapping logic and break it.

**Comment by pstibrany** [comment]
> This only makes code harder to understand and creates further bugs.

Agree. Unfortunately that's what `httpgrpc.Errorf` does today, and we use it widely in Mimir codebase.

**Comment by aknuds1** [comment]
@colega @pstibrany it should be noted though that the usage of `errorWithStatus` instead of `httpgrpc` errors was introduced before this PR (also with HTTP status codes), so it's an independent discussion.

**Comment by colega** [comment]
> Unfortunately that's what `httpgrpc.Errorf` does today, and we use it widely in Mimir codebase.

This is different from what `httpgrpc` does: `httpgrpc` creates a status with a details field that wraps an http response. Then on the caller side it is possible to distinguish a httpgrpc error by calling `httpgrpc.HTTPResponseFromError` which checks that those details are there and parses them: it doesn't just blindly assume whatever is in the `Status.Code` is a HTTP status.

With this approach there's no way to tell.

**Pull request merged**